src/plugins/plugin-redis.js

```diff
@@ -0,0 +1,210 @@
+    var client = createClient.apply(this, arguments);
+    api.wrapEmitter(client);
+    return client;
+  };
+}
+
+// Used for redis version > 2.3
+function createStreamWrap(api, create_stream) {
+  return function create_stream_trace() {
+    if (!this.stream) {
+      Object.defineProperty(this, 'stream', {
+        get: function () { return this._google_trace_stream; },
+        set: function (val) {
+          api.wrapEmitter(val);
+          this._google_trace_stream = val;
+        }
+      });
+    }
+    return create_stream.apply(this, arguments);
+  };
+}
+
+// Used for redis version <= 2.3
+function streamListenersWrap(api, install_stream_listeners) {
+  return function install_stream_listeners_trace() {
+    api.wrapEmitter(this.stream);
+    return install_stream_listeners.apply(this, arguments);
+  };
+}
+
+function setupSpan(api, cmd, args, skipped_frames) {
+  var span = api.createChildSpan({
+    name: 'redis-' + cmd,
+    skipFrames: skipped_frames + 1
+  });
+  span.addLabel('command', cmd);
+  if (api.enhancedDatabaseReportingEnabled()) {
+    span.addLabel('arguments', JSON.stringify(args));
+  }
+  return span;
+}
+
+function startSpanFromArguments(api, cmd, args, cb, send_command) {
+  if (!cmd || !args || typeof cmd !== 'string' || !Array.isArray(args) ||
+    (cb && typeof cb !== 'function')) {
+    return send_command(cmd, args, cb);
+  }
+  if (!cb) {
+    if (typeof args[args.length - 1] === 'function' ||
+        typeof args[args.length - 1] === 'undefined') {
+      cb = args.pop();
+    }
+  }
+  var span = setupSpan(api, cmd, args, 1);
+  return send_command(cmd, args, wrapCallback(api, span, cb));
+}
+
+function internalSendCommandWrap(api, internal_send_command) {
+  return function internal_send_command_trace(cmd, args, cb) {
+    var root = api.getRootSpan();
+    if (!root) {
+      return internal_send_command.call(this, cmd, args, cb);
+    }
+    if (arguments.length === 1 && typeof cmd === 'object') {
+      var span = setupSpan(api, cmd.command, cmd.args, 0);
+      cmd.callback = wrapCallback(api, span, cmd.callback);
+      return internal_send_command.call(this, cmd);
+    }
+    return startSpanFromArguments(api, cmd, args, cb, internal_send_command.bind(this));
+  };
+}
+
+function sendCommandWrap(api, send_command) {
+  return function send_command_trace(cmd, args, cb) {
+    var root = api.getRootSpan();
+    if (!root) {
+      return send_command.call(this, cmd, args, cb);
+    }
+    return startSpanFromArguments(api, cmd, args, cb, send_command.bind(this));
+  };
+}
+
+function wrapCallback(api, span, done) {
+  var fn = function(err, res) {
+    if (api.enhancedDatabaseReportingEnabled()) {
+      if (err) {
+        span.addLabel('error', err);
+      }
+      if (res) {
+        span.addLabel('result', res);
+      }
+    }
+    span.endSpan();
+    if (done) {
+      done(err, res);
+    }
+  };
+  return api.wrap(fn);
+}
+
+function wrapInternalSendCommand(redis, api) {
```

Comment:
Personally, I don't find that pulling single statements into their own functions makes things more readable but it's just a matter of preference.